py/src/nupic/tests/encoders/date_test.py

```diff
@@ -0,0 +1,223 @@
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testTime(self):
+    """ Creating date encoder instance. """
+    # 3 bits for season (x4 seasons -> 12 bits), no other fields encoded
+    enc = DateEncoder(timeOfDay=5)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+
+    # Time of day has radius of 4 hours and w of 5,
+    # so each bit = 240/5min = 48min. 
+    # 14:55 is minute 14*60 + 55 = 895; 895/48 = cetral bit= bit 18.6 ->18 #FIXME we should round this, so becomes bit19
+    # We activate our 5(=w)bits around the central bit -> thus [18-(5-1)/2, 18+(5-1)/2] = [16, 20]
+    # Encoder should be 30 bits total (30 * 48 minutes = 24 hours)
+    timeOfDayExpected = (
+      [0,0,0,0,0,0,0,0,0,0,
+       0,0,0,0,0,0,1,1,1,1,
+       1,0,0,0,0,0,0,0,0,0])
+
+    expected = timeOfDayExpected
+    self.assertEqual(bits.size, 30)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testMissingValues(self):
+    """ Missing values. """
+    e = DateEncoder(timeOfDay=5)
+    mvOutput = e.encode(None)
+    self.assertEqual(sum(mvOutput.sparse), 0)
+    mvOutput = e.encode(float('nan'))
+    self.assertEqual(sum(mvOutput.sparse), 0)
+
+
+  def testHoliday(self):
+    """ Look at holiday more carefully because of the smooth transition. """
+    e = DateEncoder(holiday=5)
+    holiday    = [0,0,0,0,0,1,1,1,1,1]
+    notholiday = [1,1,1,1,1,0,0,0,0,0]
+    holiday2   = [0,0,0,1,1,1,1,1,0,0]
+
+    d = datetime.datetime(2010, 12, 25, 4, 55) #Christmas day 25th Dec, a default holiday
+    assert(all( e.encode(d).dense == holiday ))
+
+    d = datetime.datetime(2008, 12, 27, 4, 55) #12/27 is not a holiday
+    assert(all( e.encode(d).dense == notholiday ))
+
+    d = datetime.datetime(1999, 12, 26, 8, 00) #day after holiday, approaching
+    assert(all( e.encode(d).dense == holiday2 ))
+
+    d = datetime.datetime(2011, 12, 24, 16, 00) #day before holiday, approaching
+    assert(all( e.encode(d).dense == holiday2 ))
+
+
+  def testHolidayMultiple(self):
+    """ Look at holiday more carefully because of the smooth transition. """
+    e = DateEncoder(holiday=5, holidays=[(12, 25), (2018, 4, 1), (2017, 4, 16)])
+    holiday    = [0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
+    notholiday = [1, 1, 1, 1, 1, 0, 0, 0, 0, 0]
+
+    d = datetime.datetime(2011, 12, 25, 4, 55)
+    assert(all( e.encode(d).dense == holiday ))
+
+    d = datetime.datetime(2007, 12, 2, 4, 55)
+    assert(all( e.encode(d).dense == notholiday ))
+
+    d = datetime.datetime(2018, 4, 1, 16, 10)
+    assert(all( e.encode(d).dense == holiday ))
+
+    d = datetime.datetime(2017, 4, 16, 16, 10)
+    assert(all( e.encode(d).dense == holiday ))
+
+
+  def testWeekend(self):
+    """ Test weekend encoder. """
+    e  = DateEncoder(customDays=(21, ["sat", "sun", "fri"]))
+    e2 = DateEncoder(weekend=21)
```

Comment:
can you explain how custom days work? From the code I'd expect this should be (sat+sun) == (weekend==true), why Friday?

Reply:
They consider Friday night as part of the weekend.

Reply:
> They consider Friday night as part of the weekend.

oh. Another example how overspecified the encoder is. bummer

Reply:
I think this could be a useful feature.  The point of this encoder is for use in systems which analyses human behavior, and people do act differently on week-nights and weekend-nights.

IIRC this encoder was used for predicting taxi usage, which probably changed on friday night.

Reply:
> IIRC this encoder was used for predicting taxi usage, which probably changed on friday night.

OT but yes, there are problems where this is useful. But more confirmation that instead of "weekend", it should only have the customDays. 

>  human behavior,

human behavior "encoder" is an interesting topic, but more for a forum discussion

Reply:
Anyways, we should add a comment that explains Fri 18pm > is weekend, that could be unxepected.

Reply:
Definately, this should be mentioned in the docstring at the top of the class where it will be visible to the users.